### DISCUSSION TITLE
Cascade delete

### DISCUSSION THREAD

**Discussion by wojciechkrol** [payloadcms/payload#655] [Q&A] [answered]
Is it possible to delete a record with all his relations (cascade delete) without using hooks to delete each related record?

**Comment by jmikrut** [comment, marked as answer]
Hey @wojciechkrol — yes, this is possible! You'll need to write some hooks to do it. You could write a hook that can be used to traverse your collection schema and find any `relationship` fields, and for each one, you'd read the value of the relationship field and then delete all related documents using Payload's local API.

Know what I mean?

**Reply by jmikrut** [reply]
Sorry, I didn't see that you specified `without hooks`. Honestly though I think that's probably the best way to do this. It could even be wrapped up as a neat little plugin.

We don't have any use for this at the moment, but if one of our own projects needs it, we'll certainly build it into a plugin and make it public. If you do so, please share it! It shouldn't be too hard to build.

**Comment by wojciechkrol** [comment]
Hej @jmikrut,

Thank you for your answer. I will try to make a universal hook for this and let you know if I was successful :)

**Comment by bencun** [comment]
This would be extremely helpful to have as an option to enable on a per-collection basis where a collection would then be able check for any relationship type fields referencing itself when deleting an item.
While I do understand a hook could resolve this having a built-in solution should, in my opinion, be of a very high priority since it would prevent end-users from breaking stuff (e.g. broken media, broken links etc).

**Reply by DanRibbens** [reply]
I started working on this then I got covid and my wife had a baby.

https://github.com/payloadcms/payload/pull/1209

If somebody wants to fork off my PR, this is up for grabs. Otherwise I'll finish it after finishing the admin UI translations.

**Reply by juanpablob** [reply]
Any update on this? :D

**Reply by rafalkrupinski** [reply]
@juanpablob https://github.com/payloadcms/payload/pull/16246

**Comment by rafalkrupinski** [comment]
Hi, 
a search engine brought me here.

Any chance this is solved with actual cascade delete by now?

**Reply by RobinvdGriend** [reply]
It still hasn't been solved and the issue is still open: https://github.com/payloadcms/payload/issues/11177.

**Reply by rafalkrupinski** [reply]
how is that related?

**Reply by RobinvdGriend** [reply]
Excuse me, I copied the wrong issue. I've edited my initial comment to point to the relevant issue.

**Reply by rafalkrupinski** [reply]
Would you like to test it with #16246? It's just passing cascade option to drizzle

**Comment by maxsynnott** [comment]
Hi @jmikrut,

Is there any plan to prioritise this?

I'm assuming the current issue with it is that it wouldn't trigger before/after delete hooks.

However this feels like pretty core functionality for relationships and would certainly be appreciated.

**Comment by joan-teriihoania** [comment]
Linking this here: https://github.com/payloadcms/payload/issues/11177#issuecomment-5181228068

> We've tried tackling this as well. For now, we've tagged it as EXPERIMENTAL but we've already made use of it and so far, we didn't get any problems. The approach is fundamentally the same as @ mariovalney and his repository, with a bunch of utilities surrounding it. The plugin is broader and focuses on building a "graph" of all collection relationships, and the cascade deletion only makes use of the aggregated relationships.
> 
> The flow is pretty simple:
> 
> At startup, scan through the collection for relationship fields (including upload fields)
> Then, add beforeDelete hooks in order of "ascendance" to emulate the cascading effect by going through the relationship tree
> The plugin also allows for unlinking which nulls the value instead of deleting the record (kind of like ON DELETE SET NULL in SQL) or to a provided value.
> 
> If anyone wants to give it a try and see if it works properly in their own set ups, that would be nice! I will be doing so on my side and once I feel like I have something stable, I'll remove the EXPERIMENTAL tag.
> 
> https://www.npmjs.com/package/@lizardglobal/payload-collection-references
> https://github.com/LizardGlobalGH/payload-collection-references
> https://joan-teriihoania.fr/posts/payloadcms---how-to-set-up-on-delete-cascade-in-collections
> I am open to feedback! I really wanted to contribute to PayloadCMS. It's been a while since I've enjoyed learning a new framework ^^ Please be gentle 🙏 Thank you!
> 
> EDIT: If need be, you can also reach out on the PayloadCMS's discord by pinging me @joprocorp
